internal/util/context.go

```diff
@@ -0,0 +1,26 @@
+package util
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+)
+
+type contextKey string
+
+// Context keys.
+const (
+	contextKeyLogger = contextKey("logger")
+)
+
+// ContextWithLogger returns a copy of the parent context that includes the
+// logger.
+func ContextWithLogger(ctx context.Context, logger *logrus.Logger) context.Context {
+	return context.WithValue(ctx, contextKeyLogger, logger)
+}
+
+// LoggerFromContext returns the logger from the context.
+func LoggerFromContext(ctx context.Context) *logrus.Logger {
+	logger := ctx.Value(contextKeyLogger).(*logrus.Logger)
```

Comment:
so we need to imagine nil scenario for this...

it's likely we'll make mistakes and send ctx without a logger.

so for nil, we should ideally return the default singleton logger from the logrus pkg, or something similar (that doesn't have fields etc)...

it would be easy to spot those stmts as they wouldn't have any fields. and we'd pass the proper ctx in that case.

Reply:
Something I just noticed is that if we have something like
```
lg := logger.WithField("field", "value").Logger
ctx = util.ContextWithLogger(ctx, lg)
```

and

```
lg := util.LoggerFromContext(ctx)
```

the field is no respected. It looks like when doing `.Logger`, every field goes away. This is exactly the issue you had when you demoed logrus to me (if you remember).

The way to conserve the fields is by stroring `logrus.Entry` instead of `logrus.Logger`. Should we go with that one instead?

Reply:
Mmm... Using `logrus.Entry` might also be a problem since there might be cases in which we don't want to specify fields. In those cases we cannot retrieve a `logrus.Entry` unless we do some sort of "hack" :/

Reply:
yeah storing `logrus.Entry` internally in the context is not a problem, these libraries should be able to hide it.